kerastuner/tuners/ultraband.py

```diff
@@ -12,28 +12,283 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-"Variation of HyperBand algorithm."
 
-from ...engine import tuner as tuner_module
-from ...engine import oracle as oracle_module
+import queue
+import random
+from ..engine import tuner as tuner_module
+from ..engine import oracle as oracle_module
 
 
 class UltraBandOracle(oracle_module.Oracle):
-    # TODO
-    pass
+    """Oracle class for UltraBand.
+
+        Args:
+        max_trials: Int. Total number of trials
+            (model configurations) to test at most.
+            Note that the oracle may interrupt the search
+            before `max_trial` models have been tested.
+        seed: Int. The random seed. If None, it would use a random number.
+        factor: Int. The factor of the change of number of epochs
```

Comment:
I think this needs a more explicit name, e.g. `num_rounds` or `num_brackets` or something. What is it called in the paper?

Reply:
In the paper it is just a symbol. no name. I think factor is an accurate name. There are many rounds. Each round the number of epochs is increased by multiplying that factor. The number of candidates is decreased by dividing the factor.